### PR TITLE
Performance optimization on SparseConcatOp.

### PR DESCRIPTION
**Pull request by lowintelligence**
Previous 'SparseConcatOp' uses a reorder-concat-reorder implementation. In the 'Reorder' function, a sort would be performed, which decides that the overall complexity would be O(NlogN). In large (maybe rare but occurred in our production env) cases, time consumption is significant.

This commit includes optimizations below:
  - A quickpath for dim-0 concatenation. The reorder phase would not be done, which reaches O(N) complexity.
  - Parallel implementation for concatenating and sorting phases, with a parallel quick sort implementation with Eigen threadpool.
  - A naive dimension comparator used for sorting.

Here is the performance comparison on my Mac Pro (i7-4770HQ @ 2.2GHz 4c8T, 16GB DDR3 1600) with the benchmarks in this commit:
The benchmark uses 3D string tensor, the numbers 'BM_SparseConcat_X_Y_Z_T' in the the benchmark indicate 'X: valid points of each sparse; Y: number of input tensors; Z: concat dimension; T: number of threads'.
Old:
image
New:
image

### CODE REVIEW

**Comment by lowintelligence** [comment]
Any reviewer has time to take a look about this commit?  @caisq thx

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by facaiy** [comment]
@ebrevdo @gunan @josh11b Could you take a look or reassign?

**Comment by tensorflowbutler** [comment]
Nagging Reviewers @gunan, @josh11b, @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewers @gunan, @josh11b, @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

File: tensorflow/core/kernels/sparse_concat_op.cc [hunk 1]
```diff
@@ -99,54 +107,166 @@ class SparseConcatOp : public OpKernel {
...
+    // In this implementation, the concat process has 3 steps:
+    // 1. Copy all tensor indices and values to an output with the input order,
+    // and re-calculate the indices after concatation.
+    // 2. Perform a sort on indices to get the final position of each element.
+    // 3. Copy the indices and values to the final output tensor according to
+    // the sorting result.
+    Tensor output_ix(DT_INT64, TensorShape({num_entries, input_rank}));
+    Tensor output_vals(DataTypeToEnum<T>::v(), TensorShape({num_entries}));
+    Tensor output_shape(DT_INT64, TensorShape({input_rank}));
+    auto output_shape_t = output_shape.vec<int64>();
```
Comment (shahzadlone) [inline tensorflow/core/kernels/sparse_concat_op.cc]: perhaps ```const auto```?

File: tensorflow/core/kernels/sparse_concat_op.cc [hunk 2]
```diff
@@ -99,54 +107,166 @@ class SparseConcatOp : public OpKernel {
...
+    std::copy_n(out_shape_vec.begin(), input_rank, output_shape_t.data());
+
+    auto worker_threads = context->device()->tensorflow_cpu_worker_threads();
+    auto workers = worker_threads->workers;
```
Comment (shahzadlone) [inline tensorflow/core/kernels/sparse_concat_op.cc]: could these be references?

Comment (lowintelligence) [reply tensorflow/core/kernels/sparse_concat_op.cc]: Well, fixed with explicit types with reference. Thx.

File: tensorflow/core/kernels/sparse_concat_op.cc (hunk 1 above)
Comment (lowintelligence) [reply tensorflow/core/kernels/sparse_concat_op.cc]: Sure.

**Comment by lowintelligence** [comment]
@shahzadlone , any further comments? Thanks.

**Comment by penpornk** [comment]
I'm so sorry for our delay. I will review this PR soon!

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by rthadur** [comment]
@penpornk gentle ping to review this PR , @lowintelligence can you please resolve conflicts.

File: tensorflow/core/util/sorter.h [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  template <typename T, class Compare>
+  void QSort(std::vector<T>& x, const Compare& comp) {
+    if (x.size() <= 1) return;
+    if (nthreads_ <= 1 || threadpool_ == nullptr) {
+      std::sort(x.begin(), x.end(), comp);
+      return;
+    }
+
+#define HALF_L1_CACHE_SIZE 16384
+    const int block_size = HALF_L1_CACHE_SIZE / sizeof(T);
```
Comment (penpornk) [inline tensorflow/core/util/sorter.h]: We can read the L1 cache size through Eigen's helper function. See the file below for example,
https://github.com/tensorflow/tensorflow/blob/d9154e8d341746032d42e3a9080f7bb6e22a4aeb/tensorflow/core/grappler/clusters/utils.cc#L56

File: tensorflow/core/util/sorter.h [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+namespace tensorflow {
+
+namespace {
+
+int divup(int a, int b) { return (a + (b - 1)) / b; }
+
+// Recursively re-order x, which makes all elements within [first, pos) less
+// than those in [pos, last).
+template <typename T, class Compare>
+void DivideAtPos(std::vector<T>& x, Compare& comp, int first, int last,
```
Comment (penpornk) [inline tensorflow/core/util/sorter.h]: `DivideAtPost` takes `O(n)` (when `n = last-first`) and potentially more. This means this parallel QSort will still take `O(n log n)` on the critical path. How much speedup does this get over serial QSort?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you very much for the PR! I'm terribly sorry for my delay. I have a few high-level questions and some more comments.

Regarding your comments:

> First of all, this optimization for sparse concat might break the current use of util/sparse_tensor, because I don't think construct such an object from inputs and do re-order is valuable. Maybe we could discuss about whether is a better way to keep the uniform operation in util/sparse_tensor.

I think not going through sparse tensor interface is totally fine. I agree that reordering is not worth it.

> Secondly, the fast-path of concatenation on dim0 should be performed. This definitely simplify the complexity of the operation for large scale sparse features.

I agree.

> Thirdly, this work used some additional memory for index sorting, and implemented a parallel quick-sort. We could also bring out some other idea here.

I have a question about the parallel quicksort in my comment. I think additional memory usage is okay.

**Comment by gbaned** [comment]
@lowintelligence Could you please address the reviewer comments and resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tensorflowbutler** [comment]
It has been 44 days that this pull-request has stalled. Please create a new pull-request with the requested changes.

**Comment by lowintelligence** [comment]
Sorry for missing all the messages due to my work change.
@penpornk Thanks for the review comments. This PR was hanging there for a long period thus I didn't check its status very often. I'll try to fix the codes and reopen another PR later.

**Pull request closed without merging**